Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -39,57 +44,169 @@ def _break(name):
 def _tail_call():
     raise TailCall()
 
-class Seq(list):
+class Seq(tuple):
     def __init__(self, __iterable = None, isStr = False):
         if __iterable is None:
             __iterable = []
-        self.isStr = isinstance(__iterable, str) or isStr
-        super().__init__(__iterable)
+        self.isStr = isStr or isinstance(__iterable, str) or (isinstance(__iterable, Seq) and __iterable.isStr) or (not isinstance(__iterable, GeneratorType) and all(isinstance(e, str) and len(e) == 1 for e in __iterable) and len(__iterable) > 0)
 
     @property
     def Elements(self):
         return self
 
     @property
     def UniqueElements(self):
-        return set(self)
+        return frozenset(self)
 
     def __repr__(self) -> str:
         if self.isStr:
-          return ''.join(self)
-        return super().__repr__()
+            return ''.join(self)
+        return '[' + ', '.join(map(repr, self)) + ']'
 
     def __add__(self, other):
         return Seq(super().__add__(other), isStr=self.isStr and other.isStr)
 
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            indices = range(*key.indices(len(self)))
+            return Seq((self[i] for i in indices), isStr=self.isStr)
+        return super().__getitem__(key)
+
+    def set(self, key, value):
+        l = list(self)
+        l[key] = value
+        return Seq(l, isStr=self.isStr)
+
     def __hash__(self) -> int:
         return hash(tuple(self))
 
-class Set(set):
+class Set(frozenset):
     @property
     def Elements(self):
         return self
 
+    @property
+    def AllSubsets(self):
+        complete = set(self)
+        queue = deque([set()])
+        yield Set(queue[0])
+        while True:
+            s = queue.pop()
+            missing = complete - s
+            if missing == set():
+                break
+            for e in missing:
+                n = s.copy()
+                n.add(e)
+                yield Set(n)
+                queue.appendleft(n)
```

Comment:
https://docs.python.org/3/library/itertools.html#itertools-recipes has a recipe for powerset that's short and sweet

Reply:
Thanks!